firecracker/microvm.py

```diff
@@ -369,23 +369,31 @@ async def unix_client_connected(*_):
 
     async def shutdown(self):
         logger.debug(f"Shutown vm={self.vm_id}")
-        reader, writer = await asyncio.open_unix_connection(path=self.vsock_path)
-        payload = b"halt"
-        writer.write(b"CONNECT 52\n" + payload)
+        try:
+            reader, writer = await asyncio.open_unix_connection(path=self.vsock_path)
+        except (FileNotFoundError, ConnectionResetError) as error:
+            logger.warning(f"VM={self.vm_id} cannot receive shutdown signal: {error.args}")
+            return
+
+        try:
+            payload = b"halt"
+            writer.write(b"CONNECT 52\n" + payload)
 
-        await writer.drain()
+            await writer.drain()
 
-        ack: bytes = await reader.readline()
-        logger.debug(f"ack={ack.decode()}")
+            ack: bytes = await reader.readline()
+            logger.debug(f"ack={ack.decode()}")
 
-        msg: bytes = await reader.readline()
-        logger.debug(f"msg={msg}")
+            msg: bytes = await reader.readline()
```

Comment:
readline + binary output? Seems weird, and you don't use decode in the other logs so I imagine it's actually a string and the type hint is wrong.

Reply:
The type hint is correct: readline can be used on bytes arrays.